Feature/9811 open tone bottom sheet

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->
Closes: #9811
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Add select tone bottom sheet and save selected value in main viewmodel

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Click on "Add Product" button from products list and select add product with AI. The site must be WP.com to be eligible for this option. 
2. Click "Continue" in the first step. 
3. Click on "Set tone and voice" 
4. Check the UI matches the design
5. Select any tone
6. Click again on "Set tone and voice" and check your previous selected option is now selected. 

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/2663464/32357e10-5323-4c97-bbc9-d9afa459c7fe

JorgeMucientes committed:
Trigger event to navigate to tone bottom sheet

JorgeMucientes committed:
Navigate to AddProductWithAISetToneBottomSheet

JorgeMucientes committed:
Add ai tone bottom sheet UI

JorgeMucientes committed:
Fix detekt

JorgeMucientes committed:
Propagate selected ai tone from screen to bottom sheet and viceversa

JorgeMucientes committed:
Pass selected ai tone back to subviewmodel

JorgeMucientes committed:
Remove unnecessary tone field

JorgeMucientes committed:
Replace bottom sheet fragment with compose ModalBottomSheetLayout

JorgeMucientes committed:
Hide bottom sheet on item selected

JorgeMucientes committed:
Make bottom sheet content scrollable

JorgeMucientes committed:
Expand bottom sheet only halfway if device in landscape

JorgeMucientes commented on WooCommerce/src/main/res/navigation/nav_graph_main.xml:
```diff
@@ -602,8 +602,7 @@
         <action
```
Changes to this file can be ignored. Just reformatting.

JorgeMucientes committed:
Merge branch 'trunk' into feature/9811-open-tone-bottom-sheet

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AboutProductSubScreen.kt

JorgeMucientes committed:
Use correct dimen value

JorgeMucientes committed:
Handle continue click to be able to navigate to about your product step

JorgeMucientes committed:
Show placeholder always in about your product text field

JorgeMucientes committed:
Remove unnecessary events field override

JorgeMucientes committed:
Revert unnecessary code formatting

hichamboushaba commented on WooCommerce/src/main/res/navigation/nav_graph_main.xml:
```diff
@@ -602,8 +602,7 @@
             app:destination="@id/getPaidFragment"
-            app:popUpTo="@id/dashboard">
-        </action>
+            app:popUpTo="@id/dashboard"></action>
```
np, since you reformatted the file, I think it would be great to close the XML tag here using `/>` instead of the additional `</action>` which looks out of place.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AboutProductSubViewModel.kt:
```diff
@@ -1,43 +1,59 @@
+        val selectedAiTone: AiTone
+    ) : Parcelable
+
+    enum class AiTone(@StringRes val displayName: Int) {
```
Just a minor point, WDYT about adding an English name property? the one we'll pass to the AI, I know we can use the `name` property, but I feel it's better to be explicit about the value we will be passing 🤔

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AboutProductSubScreen.kt:
```diff
@@ -41,93 +59,194 @@ fun AboutProductSubScreen(viewModel: AboutProductSubViewModel, modifier: Modifie
+    selectedTone: AiTone,
+    onToneSelected: (AiTone) -> Unit
+) {
+    Surface(
+        shape = RoundedCornerShape(
+            topStart = dimensionResource(id = R.dimen.minor_100),
+            topEnd = dimensionResource(id = R.dimen.minor_100)
+        )
+    ) {
```
The `ModalBottomSheetLayout` already defines the shape of the Sheet, so we can safely remove the Surface declaration from here, and have the content directly.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AboutProductSubScreen.kt:
```diff
@@ -41,93 +59,194 @@ fun AboutProductSubScreen(viewModel: AboutProductSubViewModel, modifier: Modifie
+
+    ModalBottomSheetLayout(
+        sheetState = modalSheetState,
+        sheetShape = RoundedCornerShape(topStart = 12.dp, topEnd = 12.dp),
```
I think the value should be 8.dp, or better `dimenResource(id = R.dimen.minor_100)`
```suggestion
        sheetShape = RoundedCornerShape(
            topStart = dimensionResource(id = R.dimen.minor_100),
            topEnd = dimensionResource(id = R.dimen.minor_100)
        ),
```

hichamboushaba reviewed (commented):
Nice work @JorgeMucientes, I left some np comments, and there is a remark regarding the dark UI, I see two issues:

| View | Compose |
| --- | --- |
| ![Screenshot_20230922_163808](https://github.com/woocommerce/woocommerce-android/assets/1657201/2ef37504-d585-4b73-b125-e592e76cbb97) | ![Screenshot_20230922_163823](https://github.com/woocommerce/woocommerce-android/assets/1657201/a1df8b1c-64ec-4833-9a64-f9cad3f6e1c5) |

- One issue that's very easy to fix is the `scrim` color, the default value of compose doesn't match the value of our theme, so we need to override it:
```kotlin
scrimColor = colorResource(id = R.color.color_scrim_background)
```

- The other issue is about the background color of the sheet, I still don't understand why it happens, as we should be using the same elevation for both sheets.
Technically, the Compose version is more accessible and matches the Material specs better, but it's inconsistent with the rest of the app.

I don't think this last point is a blocker, I just mention it so that we keep it in mind, and maybe spend some time to understand the cause, and maybe discuss with the designers to fix the background for the View system, which is not accessible at all.
So once you fix the first issue, and review the other comments, this will be ready to ship 🚀.

JorgeMucientes committed:
Remove unnecessary tag

JorgeMucientes committed:
Add a slug to ai tone to avoid using enum name property

JorgeMucientes committed:
Use the correct value to set rounded corners

JorgeMucientes committed:
Fix composable bottom sheet scrim color

JorgeMucientes committed:
Merge branch 'trunk' into feature/9811-open-tone-bottom-sheet

# Conflicts:
#	WooCommerce/src/main/res/values/strings.xml

JorgeMucientes committed:
Add workaround to make composable bottom sheet scrim color work

JorgeMucientes commented:
Excellent feedback @hichamboushaba. Thank you! I've addressed all the concerns except for the elevation issue resulting in the sheet having a different background color in dark mode. 
From a quick look at it I don't understand why it happens either. 
- I checked and `BottomSheetDialogFragment` relies on `Widget.Design.BottomSheet.Modal` style which sets elevation to 16dp through `design_bottom_sheet_modal_elevation">16dp</dimen>`

- And Compose `ModalBottomSheetLayout` uses `ModalBottomSheetDefaults.Elevation` which is also 16dp

- Additionally, both components use `color_surface` as background. 

- I don't know where the darker color for the `BottomSheetDialogFragments` is coming from. Because even if I force `ModalBottomSheetLayout` to have an elevation of `1dp` the color is still brighter in `ModalBottomSheetLayout` than the rest of the sheets relying on `BottomSheetDialogFragments`

Since I wasn't able to find the cause of the problem, I left it as is. I agree we can ping Ann to see if we should improve color contrast for all sheets in the app for better accessibility. However it seems will not be straightforward to fix that, as we don't know where the dark color for `BottomSheetDialogFragments` is coming from 😅. 

PS: I had to add a workaround: https://github.com/woocommerce/woocommerce-android/pull/9816/commits/ad77cbaee0a0c74a4d81f559c4fa479fc3a1cbc0 for scrim color. It seems [there's a bug](https://issuetracker.google.com/issues/183697056) which makes the `scrim` color have no effect in light mode: 

https://github.com/woocommerce/woocommerce-android/assets/2663464/c818d88a-1c71-4876-8237-8766934896d7

I think if we update the BOM to more recent version the issue will be fixed. But that requires some migration as there are multiple breaking changes. 

Anyway I think the feature is ready for shipping now 🙂

hichamboushaba commented:
Thank you @JorgeMucientes 

> Since I wasn't able to find the cause of the problem, I left it as is.

I think that's OK, we will try to investigate this more when the right time comes.

> PS: I had to add a workaround: https://github.com/woocommerce/woocommerce-android/commit/ad77cbaee0a0c74a4d81f559c4fa479fc3a1cbc0 for scrim color. It seems [there's a bug](https://issuetracker.google.com/issues/183697056) which makes the scrim color have no effect in light mode:

Nice find, WDYT about adding a comment about this, and mentioning the bug? this way we'll remember about it.

hichamboushaba reviewed (approved):
Nice work @JorgeMucientes 👏, this is ready to ship, but I'll leave it unmerged for you to check my last comment.

JorgeMucientes committed:
Add explanatory comment to the scrim color workaround
